lib/sensors/MoistureSensor.cpp

```diff
@@ -0,0 +1,93 @@
+#include "mbed.h"
+#include "MoistureSensor.h"
+
+constexpr int Sensor_I2C_Address = 0x36 << 1;                               //MBED I2C uses 8 bit addressing, so addresses are left shifted by 1 (may need to be shifted by 2) 
+
+constexpr int Sensor_Status_Base = 0x00;                                    //Base address registers for different modules
+constexpr int Sensor_Moisture_Base = 0x0F;
+
+constexpr int Sensor_Status_HW_ID = 0x01;                                   //Function address register for the sensor's HW ID
+constexpr int Sensor_HW_ID_Code = 0x68;                                     //Expected value for sensor HW ID; NOTE: This is not the expected value as specified by the datasheet,
+                                                                            //it's just what I found through my own testing. It sometimes fluctuates on its own, so it's hard to trust :(
+
+constexpr int Sensor_Moisture_Function = 0x10;                              //Function address registers for moisture and temperature modules        
+constexpr int Sensor_Temp_Function = 0x04;  
+
+MoistureSensor::MoistureSensor(PinName sda, PinName scl) : i2c_(sda, scl){}
+
+bool MoistureSensor::Is_Initialized(){
+    if(this->Read_HW_ID() == Sensor_HW_ID_Code){                            //compare received HW ID Code to correct one
+        return true;
+    }
+
+    return false;
+}
+
+int MoistureSensor::Read_HW_ID(){
+    char cmd[2];
+    cmd[0] = Sensor_Status_Base;
+    cmd[1] = Sensor_Status_HW_ID;
+
+    char check[1];
+
+    i2c_.write(Sensor_I2C_Address, cmd, 2);                                 //initialize registers for checking device ID
+
+    i2c_.read(Sensor_I2C_Address, check, 1);                                //read device ID
+
+    return check[0];
+}
+
+uint16_t MoistureSensor::Read_Moisture(){
+    /*
+    if(!(this->Is_Initialized())){                                          //checks if device is initialized, returns 65534 if there is an issue
+        return 65534;                                                       //this is commented out because the HW_ID reading isn't consistent currently
+    }
+    */
+
+    char cmd[2];
+    cmd[0] = Sensor_Moisture_Base;
+    cmd[1] = Sensor_Moisture_Function;
+
+    char buf[2];
+    uint16_t ret = 65535;
+
+    do{
+        wait_ms(1);
```

Comment:
cant check at the moment but if wait_ms just burns cycles you should potentially use ThisThread::sleep(1) here instead (especially if in this loop for a while)

Reply:
It isn't a busy wait, but still should switch to sleep.

from [mbed docs](https://os.mbed.com/docs/mbed-os/v5.15/apis/wait.html):

> Note: wait is deprecated in favor of explicit sleep functions. To sleep, replace wait with ThisThread::sleep_for (C++) or thread_sleep_for (C). To wait (without sleeping), call wait_us. wait_us is safe to call from ISR context.